### ISSUE TITLE
tn3270 active extended key panel

### ISSUE THREAD

**Issue by ibmrcruicks** [zowe/zlux#757]
currently no way in the emulator to know fixed key-map assignments for function/PA keys, SYSREQ, EOF, ATTN, etc without refering to wiki.

A panel at the bottom of the screen below OIA with buttons that show a couple of rows with buttons for the available keys, which inject the appropriate event when pressed. 

No need for customisation since they would be related to key presses, not assigned functionality.

This would allow greater use of the Zowe desktop for 3270 as a way to avoid browser and OS variants of other emulator options.

**Comment by 1000TurquoisePogs** [comment]
I think @Martin-Zeithaml was creating some prototypes of buttons assigned to keys, so that may be a build of the terminal worth installing when he is ready.
You can actually install apps into Zowe without upgrading zowe itself, so if Martin has a build, it could be shared early.

**Comment by ibmrcruicks** [comment]
that's good news - happy to act as a trial site for a stable build

**Comment by Martin-Zeithaml** [comment]
Yes, I am working on an extension of the TN3270 terminal. The idea is more general and it is based on the customization, but it can fulfill your requirement with that.

This is the current state, there is new bar where you can use own key actions:
![image](https://user-images.githubusercontent.com/66114686/139051061-79a61097-3a09-4719-9452-60d5d73b53d0.png)

The definition of the buttons is in the JSON file, just small introduction:
* Title - will be displayed on the bar
* Description - hover help if you keep mouse cursor on the button
* Keys - the sequence of the keys you would like to emulate
* In example below:
  * HelloW - just types Hello, world!
  * SWAP NEXT - presses Home to get on the command line, Ctrl+E (clears the input field), type SWAP NEXT, hits Enter
  * Reset, PA1 and PF23 - special terminal keys defined based on the wiki
* The syntax of this might be changed
```
{   
    "title": "HelloW",
    "description": "Hello, world!",
    "keys": [
          { "normal": "Hello, world!" }
    ]
},
{
    "title": "SWAP NEXT",
    "description": "ISPF: Swap next command",
    "keys": [
        { "special": "Home" },
        { "normal": "E", "ctrl": "true" },
        { "normal": "SWAP NEXT" },
        { "special": "Enter" }
    ]
},
{
    "title": "Reset",
    "description": "Reset",
    "keys": [
        { "normal" : "r", "alt" : "true" }
    ]
},
{
    "title": "PA1",
    "description": "Attention",
    "keys": [
        { "normal" : "1", "alt" : "true" }
    ]
},
{
    "title": "PF23",
    "description": "PF23",
    "keys": [
        { "special" : "F11", "shift" : "true" }
    ]
}
```
There is (at this moment) this limitation:
* You can define and modify keys only by editing the JSON file (file is located in workspace of instance)

Note: if this goes to the production, it will be good to use all "wiki" shortcuts and make it as default JSON definition.

**Comment by ibmrcruicks** [comment]
that looks like it might be ideal for our teaching purposes - the last thing we need is thousands of customised key/button assignments.  A default button list that matches the default keys would be great.

here's a thought - two button panels - one for custom "macros", the other for AID triggers.
This kind of thing - switchable from leftside to right side for easier "handedness"
<img width="695" alt="3270-keys-zowe" src="https://user-images.githubusercontent.com/25483279/139064106-dd9e38f0-ffe0-4099-9755-a983fe57db87.png">

**Comment by ibmrcruicks** [comment]
would also work with touch-screen devices

**Comment by Martin-Zeithaml** [comment]
There are already some other ideas submitted to this TN3270 terminal - such as to be able to define and use more connections. That's why we have also started with research how to change/improve UI of the terminal to have for example menu and dialogs. It is very possible, that in the future the current feature of user defined keys will get better UI with ability to choose the position of key bar.

But the plan is to push the current implementation as it is to the production to get feedback from more users.

**Comment by ibmrcruicks** [comment]
if the current implementation will support ~33 AID keys, that would be a great interim option, and I'd give it a trial with a few of our users

**Comment by DivergentEuropeans** [comment]
@Martin-Zeithaml As discussed in our call, if relevant for good UX (say to match button size), here's an example of an app listening to theme information coming from the Desktop:

https://github.com/zowe/zlux-editor/pull/150/files
